src/tiffvisitor_int.cpp

```diff
@@ -462,6 +463,78 @@ namespace Exiv2 {
         }
     } // TiffMetadataDecoder::decodeIptc
 
+    static const TagInfo* findTag(const TagInfo* pList,uint16_t tag)
+    {
+        while ( pList->tag_ != 0xffff && pList->tag_ != tag ) pList++;
+        return  pList->tag_ != 0xffff  ? pList : NULL;
+    }
+
+    void TiffDecoder::decodeCanonAFInfo(const TiffEntryBase* object) {
+        // report Exif.Canon.AFInfo as usual
+        TiffDecoder::decodeStdTiffEntry(object);
+        if ( !object->pValue()->count() || object->pValue()->typeId() != unsignedShort ) return; // no data!
+
+        // create vector of signedShorts from unsignedShorts in Exif.Canon.AFInfo
+        std::vector<int16_t>  ints;
+        std::vector<uint16_t> uint;
+        for (int i = 0; i < object->pValue()->count(); i++) {
+            ints.push_back((int16_t) object->pValue()->toLong(i));
+            uint.push_back((uint16_t) object->pValue()->toLong(i));
+        }
+        // Check this is AFInfo2 (ints[0] = bytes in object)
+        if ( ints[0] != object->pValue()->count()*2 ) return ;
+
+        std::string familyGroup(std::string("Exif.") + groupName(object->group()) + ".");
+
+        const uint16_t nPoints = uint[2];
+        const uint16_t nMasks  = (nPoints+15)/(sizeof(uint16_t) * 8);
+        int            nStart  = 0;
+
+        struct {
+            uint16_t tag    ;
+            uint16_t size   ;
+            bool     bSigned;
+        } records[] = {
+            { 0x2600 , 1       , true  }, // AFInfoSize
+            { 0x2601 , 1       , true  }, // AFAreaMode
+            { 0x2602 , 1       , true  }, // AFNumPoints
+            { 0x2603 , 1       , true  }, // AFValidPoints
+            { 0x2604 , 1       , true  }, // AFCanonImageWidth
+            { 0x2605 , 1       , true  }, // AFCanonImageHeight
+            { 0x2606 , 1       , true  }, // AFImageWidth"
+            { 0x2607 , 1       , true  }, // AFImageHeight
+            { 0x2608 , nPoints , true  }, // AFAreaWidths
+            { 0x2609 , nPoints , true  }, // AFAreaHeights
+            { 0x260a , nPoints , true  }, // AFXPositions
+            { 0x260b , nPoints , true  }, // AFYPositions
+            { 0x260c , nMasks  , false }, // AFPointsInFocus
+            { 0x260d , nMasks  , false }, // AFPointsSelected
+            { 0x260e , nMasks  , false }, // AFPrimaryPoint
+            { 0xffff , 0       , true  }  // end marker
+        };
+        // check we have enough data!
+        uint16_t count = 0;
+        for ( uint16_t i = 0; records[i].tag != 0xffff ; i++) count += records[i].size ;
```

Comment:
Isn't this equivalent to:
```suggestion
        const uint16_t count = 8 + 4 * nPoints + 3 * nMasks;
```

Reply:
Well yes, and no.  It's not static.  At the moment it's 4 records of nPoints and 3 of nMasks.  However In development, we've modified the number of records with nPoints and nMasks in the `records[]` array more than once.  It's safer to count how much data we need, and be certain it's available.